php-packages/phpstan/src/Attributes/ModelCastAttributeExtension.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\PHPStan\Attributes;
+
+use Carbon\Carbon;
+use Flarum\PHPStan\Extender\MethodCall;
+use Flarum\PHPStan\Extender\Resolver;
+use PHPStan\PhpDoc\TypeStringResolver;
+use PHPStan\Reflection\ClassReflection;
+use PHPStan\Reflection\PropertiesClassReflectionExtension;
+use PHPStan\Reflection\PropertyReflection;
+use PHPStan\Type\NullType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\UnionType;
+
+class ModelCastAttributeExtension implements PropertiesClassReflectionExtension
+{
+    /** @var Resolver */
+    private $extendersResolver;
+    /** @var TypeStringResolver */
+    private $typeStringResolver;
+
+    public function __construct(Resolver $extendersResolver, TypeStringResolver $typeStringResolver)
+    {
+        $this->extendersResolver = $extendersResolver;
+        $this->typeStringResolver = $typeStringResolver;
+    }
+
+    public function hasProperty(ClassReflection $classReflection, string $propertyName): bool
+    {
+        return $this->findCastAttributeMethod($classReflection, $propertyName) !== null;
+    }
+
+    public function getProperty(ClassReflection $classReflection, string $propertyName): PropertyReflection
+    {
+        return $this->resolveCastAttributeProperty($this->findCastAttributeMethod($classReflection, $propertyName), $classReflection);
+    }
+
+    private function findCastAttributeMethod(ClassReflection $classReflection, string $propertyName): ?MethodCall
+    {
+        foreach ($this->extendersResolver->getExtenders() as $extender) {
+            if (! $extender->isExtender('Model')) {
+                continue;
+            }
+
+            foreach (array_merge([$classReflection->getName()], $classReflection->getParentClassesNames()) as $className) {
+                if ($className === 'Flarum\Database\AbstractModel') {
```

Comment:
Same here, might be cleaner to put all these strings in one place. Deriving it from the class would be nice, but I assume impossible since then phpstan ext would need to depend on core